File: app/src/main/java/sgtmelon/scriptum/app/model/item/StatusItem.java
package sgtmelon.scriptum.app.model.item;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.os.Build;

import java.util.List;

import androidx.core.app.NotificationCompat;
import androidx.core.app.TaskStackBuilder;
import sgtmelon.scriptum.R;
import sgtmelon.scriptum.app.database.RoomDb;
import sgtmelon.scriptum.app.model.NoteRepo;
import sgtmelon.scriptum.app.view.activity.SplashActivity;
import sgtmelon.scriptum.office.utils.ColorUtils;

/**
 * Управление закреплением заметки в статус баре {@link NoteRepo}
 */
public final class StatusItem {

    private final Context context;
    private final PendingIntent pendingIntent;

    private NoteItem noteItem;
    private Notification notification;
    private NotificationManager notificationManager;

    // TODO: 29.08.2018 Добавить кнопки к уведомлениям, чтобы была возможность их открепить

    public StatusItem(Context context, NoteItem noteItem, boolean notify) {
        this.context = context;

        final TaskStackBuilder stackBuilder = TaskStackBuilder.create(context);
        stackBuilder.addNextIntent(SplashActivity.Companion.getInstance(context, noteItem.getId()));

        pendingIntent = stackBuilder.getPendingIntent((int) noteItem.getId(), PendingIntent.FLAG_UPDATE_CURRENT);

        updateNote(noteItem, notify);
    }

    /**
     * Обновление информации о заметке
     *
     * @param noteItem - Модель заметки
     * @param notify   - Необходимость обновления
     */
    public void updateNote(NoteItem noteItem, boolean notify) {
        this.noteItem = noteItem;

        int icon = 0;
        String text = "";

        switch (noteItem.getType()) {
            case TEXT:
                icon = R.drawable.notif_bind_text;
                text = noteItem.getText();
                break;
            case ROLL:
                icon = R.drawable.notif_bind_roll;

                final RoomDb db = RoomDb.provideDb(context);
                text = db.daoRoll().getText(noteItem.getId(), noteItem.getText());
                db.close();
                break;
        }

        final NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(context, context.getString(R.string.notification_channel_id))
                .setSmallIcon(icon)
                .setColor(ColorUtils.INSTANCE.get(context, noteItem.getColor(), true))
                .setContentTitle(noteItem.getName(context))
                .setContentText(text)
                .setCategory(NotificationCompat.CATEGORY_EVENT)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(text))
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setContentIntent(pendingIntent)
                .setAutoCancel(false)
                .setOngoing(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            notificationBuilder.setGroup(context.getString(R.string.notification_group));
        }

        notification = notificationBuilder.build();
        notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            final NotificationChannel notificationChannel = new NotificationChannel(context.getString(R.string.notification_channel_id), context.getString(R.string.notification_channel), NotificationManager.IMPORTANCE_DEFAULT);
            notificationChannel.setSound(null, null);
            notificationChannel.setVibrationPattern(null);

            notificationManager.createNotificationChannel(notificationChannel);
        }

        if (noteItem.isStatus() && notify) {
            notifyNote();
        }
    }

    /**
     * В окне редактирования заметок
     *
     * @param noteItem  - Модель заметки
     * @param rkVisible - Видимые категории
     */
    public void updateNote(NoteItem noteItem, List<Long> rkVisible) {
        if (!noteItem.isStatus()) return;

        final List<Long> rankId = noteItem.getRankId();
        if (rankId.size() == 0 || rkVisible.contains(rankId.get(0))) {
            updateNote(noteItem, true);
        } else {
            cancelNote();
        }
    }

    /**
     * Показывает созданное уведомление
     */
    public void notifyNote() {
        notificationManager.notify((int) noteItem.getId(), notification);
    }

    /**
     * Убирает созданное уведомление
     */
    public void cancelNote() {
        notificationManager.cancel((int) noteItem.getId());
    }

}

File: app/src/main/java/sgtmelon/scriptum/app/model/item/StatusItem.kt
package sgtmelon.scriptum.app.model.item

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.TaskStackBuilder
import sgtmelon.scriptum.R
import sgtmelon.scriptum.app.model.NoteModel
import sgtmelon.scriptum.app.model.key.NoteType
import sgtmelon.scriptum.app.repository.RoomRepo
import sgtmelon.scriptum.app.screen.view.SplashActivity
import sgtmelon.scriptum.office.utils.ColorUtils

/**
 * Управление закреплением заметки в статус баре [NoteModel]
 *
 * @author SerjantArbuz
 */
class StatusItem(private val context: Context, noteItem: NoteItem, notify: Boolean) {

    // TODO: 29.08.2018 Добавить кнопки к уведомлениям, чтобы была возможность их открепить

    private val pendingIntent: PendingIntent? = TaskStackBuilder.create(context).apply {
        addNextIntent(SplashActivity.getIntent(context, noteItem.type, noteItem.id))
    }.getPendingIntent(noteItem.id.toInt(), PendingIntent.FLAG_UPDATE_CURRENT)

    private lateinit var noteItem: NoteItem

    private lateinit var notification: Notification
    private val notificationManager =
            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    init {
        updateNote(noteItem, notify)
    }

    /**
     * Обновление информации о заметке, [notify] - обновление в статус баре
     */
    fun updateNote(noteItem: NoteItem, notify: Boolean = true) {
        this.noteItem = noteItem

        var icon = 0
        var text = ""

        when (noteItem.type) {
            NoteType.TEXT -> {
                icon = R.drawable.notif_bind_text
                text = noteItem.text
            }
            NoteType.ROLL -> {
                icon = R.drawable.notif_bind_roll
                text = RoomRepo.getInstance(context).getRollStatusString(noteItem) //  TODO !!
            }
        }

        val notificationBuilder = NotificationCompat.Builder(context, context.getString(R.string.notification_channel_id))
                .setSmallIcon(icon)
                .setColor(ColorUtils.get(context, noteItem.color, true))
                .setContentTitle(noteItem.getStatusName(context))
                .setContentText(text)
                .setCategory(NotificationCompat.CATEGORY_EVENT)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setStyle(NotificationCompat.BigTextStyle().bigText(text))
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setContentIntent(pendingIntent)
                .setAutoCancel(false)
                .setOngoing(true)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            notificationBuilder.setGroup(context.getString(R.string.notification_group))
        }

        notification = notificationBuilder.build()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationManager.createNotificationChannel(NotificationChannel(
                    context.getString(R.string.notification_channel_id), context.getString(R.string.notification_channel), NotificationManager.IMPORTANCE_DEFAULT
            ).apply {
                setSound(null, null)
                vibrationPattern = null
            })
        }

        if (noteItem.isStatus && notify) {
            notifyNote()
        }
    }

    /**
     * В окне редактирования заметок, [listRankVisible] - id видимых категорий
     */
    fun updateNote(noteItem: NoteItem, listRankVisible: List<Long>) {
        if (!noteItem.isStatus) return

        val rankId = noteItem.rankId
        if (rankId.isEmpty() || listRankVisible.contains(rankId[0])) {
            updateNote(noteItem)
        } else {
            cancelNote()
        }
    }

    /**
     * Показывает созданное уведомление
     */
    fun notifyNote() = notificationManager.notify(noteItem.id.toInt(), notification)

    /**
     * Убирает созданное уведомление
     */
    fun cancelNote() = notificationManager.cancel(noteItem.id.toInt())

}
